pennylane/gradients/pulse_generator_gradient.py

```diff
@@ -0,0 +1,578 @@
+
+    .. note::
+
+        Currently, this function work via tensor multiplication, costing
+        :math:`\mathcal{O}(n16^N)` scalar multiplications, where :math:`N` is the
+        number of qubits the evolution acts on and :math:`n` is the number of (scalar) parameters
+        of the evolution. This can be improved to :math:`\mathcal{O}(nN4^N}` by using the
+        fast Walsh-Hadamard transform.
+    """
+    return tuple(qml.math.real(2j * _pauli_decompose(g, num_wires)) for g in generators)
+
+
+def _nonzero_coeffs_and_words(coefficients, num_wires, atol=1e-8):
+    """Given a tuple of coefficient tensors with a common first axis size :math:`4^N-1`,
+    where :math:`N` is the number of wires, filter out those indices for the first axis for which
+    the sliced coefficients are not all zero. Return these coefficients, as well as the
+    corresponding Pauli word strings on :math:`N` wires for these indices.
+
+    Args:
+        coefficients(tuple[tensor_like]): Coefficients to filter.
+        num_wires (int): Number of qubits the generators act on.
+        atol (float): absolute tolerance used to determine whether a tensor is zero.
+    """
+    words = pauli_basis_strings(num_wires)
+    new_coefficients = [[] for _ in coefficients]
+    new_words = []
+    for word, *coeffs in zip(words, *coefficients):
+        if all(qml.math.allclose(c, 0.0, atol=atol, rtol=0.0) for c in coeffs):
+            continue
+        new_words.append(word)
+        for new_coeffs, c in zip(new_coefficients, coeffs):
+            new_coeffs.append(c)
+    return new_coefficients, new_words
+
+
+def _insert_op(tape, ops, op_idx):
+    """Create new tapes, each with an additional operation from a provided list inserted at
+    the specified position. Creates as many new tapes as there are operations provided.
+
+    Args:
+        tape (`~.QuantumTape`): Original tape.
+        ops (list[qml.Operation]): Operations to insert (individually) at ``op_idx``
+            to produce a new tape each.
+        op_idx (int): Index at which to insert the operations in ``ops``.
+
+    Returns:
+        list[`~.QuantumScript`]: new tapes with inserted operations, one tape per operation
+        in ``ops``.
+    """
+    ops_pre = tape.operations[:op_idx]
+    ops_post = tape.operations[op_idx:]
+    return [
+        qml.tape.QuantumScript(ops_pre + [insert] + ops_post, tape.measurements) for insert in ops
+    ]
+
+
+def _generate_tapes_and_coeffs(tape, idx, atol, cache):
+    """Compute the modified tapes and coefficients required to compute the pulse generator
+    derivative of a tape with respect to an indicated trainable parameter.
+
+    Args:
+        tape (`~.QuantumTape`): Tape to differentiate.
+        idx (int): Index of the parameter with respect to which to differentiate
+            into ``tape.trainable_parameters``.
+        atol (float): absolute tolerance used to determine whether a coefficient is zero.
```

Comment:
is this hardcoded or could a user provide this?

Reply:
No, this can definitely be a user input :) 
it can just be provided to `pulse_generator(tape/qnode, ... atol=...)` or in a QNode definition:
`@qml.qnode(dev, diff_method=qml.gradients.pulse_generator, atol=1e-5)`